### PR TITLE
Add modifier state to the split keyboard transport

### PR DESCRIPTION
**Pull request by cwebster2**
## Description

This adds modifier state to the i2c and serial transport for split
keyboards.  The purpose of this is to allow e.g. displaying modifier
state on the slave side of a split keyboard on an oled.  This adds two
or three bytes to the data transferred between split halves.

This also fixes a missing ifdef guard for BACKLIGHT_ENABLE.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/split_common/transport.c [hunk 1]
```diff
@@ -28,7 +28,9 @@ static pin_t encoders_pad[] = ENCODERS_PAD_A;
 
 typedef struct _I2C_slave_buffer_t {
     matrix_row_t smatrix[ROWS_PER_HAND];
+#   if defined(BACKLIGHT_ENABLE)
```
Comment (fauxpark) [inline quantum/split_common/transport.c:31]:
```suggestion
    uint8_t real_mods;
    uint8_t weak_mods;
#ifndef NO_ACTION_ONESHOT
    uint8_t oneshot_mods;
#endif
#   ifdef BACKLIGHT_ENABLE
```

File: quantum/split_common/transport.c [hunk 2]
```diff
@@ -38,6 +40,7 @@ typedef struct _I2C_slave_buffer_t {
 #    ifdef WPM_ENABLE
     uint8_t current_wpm;
 #    endif
+    uint8_t modifiers;
```
Comment (fauxpark) [inline quantum/split_common/transport.c:43]:
```suggestion
```

File: quantum/split_common/transport.c [hunk 3]
```diff
@@ -47,6 +50,7 @@ static I2C_slave_buffer_t *const i2c_buffer = (I2C_slave_buffer_t *)i2c_slave_re
 #    define I2C_KEYMAP_START offsetof(I2C_slave_buffer_t, smatrix)
 #    define I2C_ENCODER_START offsetof(I2C_slave_buffer_t, encoder_state)
 #    define I2C_WPM_START offsetof(I2C_slave_buffer_t, current_wpm)
+#    define I2C_MODS_START offsetof(I2C_slave_buffer_t, modifiers)
```
Comment (fauxpark) [inline quantum/split_common/transport.c:53]:
```suggestion
#    define I2C_REAL_MODS_START offsetof(I2C_slave_buffer_t, real_mods)
#    define I2C_WEAK_MODS_START offsetof(I2C_slave_buffer_t, weak_mods)
#    define I2C_ONESHOT_MODS_START offsetof(I2C_slave_buffer_t, oneshot_mods)
#    define I2C_ENCODER_START offsetof(I2C_slave_buffer_t, encoder_state)
#    define I2C_WPM_START offsetof(I2C_slave_buffer_t, current_wpm)
```

File: quantum/split_common/transport.c [hunk 4]
```diff
@@ -91,6 +95,11 @@ bool transport_master(matrix_row_t matrix[]) {
         }
     }
 #    endif
+    uint8_t modifiers = get_mods() | get_weak_mods();
+    if (modifiers != i2c_buffer->modifiers) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_MODS_START, (void *)&current_wpm, sizeof(modifiers), TIMEOUT) >= 0) {
+            i2c_buffer->modifiers = modifiers;
+    }
```
Comment (fauxpark) [inline quantum/split_common/transport.c:102]:
```suggestion
    uint8_t real_mods = get_mods();
    if (real_mods != i2c_buffer->real_mods) {
        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_REAL_MODS_START, (void *)&real_mods, sizeof(real_mods), TIMEOUT) >= 0) {
            i2c_buffer->real_mods = real_mods;
    }
    uint8_t weak_mods = get_weak_mods();
    if (weak_mods != i2c_buffer->weak_mods) {
        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_WEAK_MODS_START, (void *)&weak_mods, sizeof(weak_mods), TIMEOUT) >= 0) {
            i2c_buffer->weak_mods = weak_mods;
    }
#ifndef NO_ACTION_ONESHOT
    uint8_t oneshot_mods = get_oneshot_mods();
    if (oneshot_mods != i2c_buffer->oneshot_mods) {
        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_ONESHOT_MODS_START, (void *)&oneshot_mods, sizeof(oneshot_mods), TIMEOUT) >= 0) {
            i2c_buffer->oneshot_mods = oneshot_mods;
    }
#endif
```

File: quantum/split_common/transport.c [hunk 5]
```diff
@@ -118,6 +127,8 @@ void transport_slave(matrix_row_t matrix[]) {
 #    ifdef WPM_ENABLE
     set_current_wpm(i2c_buffer->current_wpm);
 #    endif
+
+    set_mods(i2c_buffer->modifiers);
```
Comment (fauxpark) [inline quantum/split_common/transport.c:131]:
```suggestion
    set_mods(i2c_buffer->real_mods);
    set_weak_mods(i2c_buffer->weak_mods);
#ifndef NO_ACTION_ONESHOT
    set_oneshot_mods(i2c_buffer->oneshot_mods);
#endif
```

File: quantum/split_common/transport.c [hunk 6]
```diff
@@ -251,6 +263,7 @@ bool transport_master(matrix_row_t matrix[]) {
     // Write wpm to slave
     serial_m2s_buffer.current_wpm = get_current_wpm();
 #    endif
+    serial_m2s_buffer.modifiers = get_mods() | get_weak_mods();
```
Comment (fauxpark) [inline quantum/split_common/transport.c:266]:
```suggestion
    serial_m2s_buffer.real_mods = get_mods();
    serial_m2s_buffer.weak_mods = get_weak_mods();
#    ifndef NO_ACTION_ONESHOT
    serial_m2s_buffer.oneshot_mods = get_oneshot_mods();
#    endif
```

File: quantum/split_common/transport.c [hunk 7]
```diff
@@ -271,6 +284,7 @@ void transport_slave(matrix_row_t matrix[]) {
 #    ifdef WPM_ENABLE
     set_current_wpm(serial_m2s_buffer.current_wpm);
 #    endif
+    set_mods(serial_m2s_buffer.modifiers);
```
Comment (fauxpark) [inline quantum/split_common/transport.c:287]:
```suggestion
    set_mods(serial_m2s_buffer.real_mods);
    set_weak_mods(serial_m2s_buffer.weak_mods);
#    ifndef NO_ACTION_ONESHOT
    set_oneshot_mods(serial_m2s_buffer.oneshot_mods);
#    endif
```

File: quantum/split_common/transport.c [hunk 8]
```diff
@@ -145,6 +156,7 @@ typedef struct _Serial_m2s_buffer_t {
 #    ifdef WPM_ENABLE
     uint8_t current_wpm;
 #    endif
+    uint8_t modifiers;
```
Comment (fauxpark) [inline quantum/split_common/transport.c:159]:
```suggestion
    uint8_t real_mods;
    uint8_t weak_mods;
#    ifndef NO_ACTION_ONESHOT
    uint8_t oneshot_mods;
#    endif
```
And this should be moved up above `backlight_level`.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/split_common/transport.c (hunk 1 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 2 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 3 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 4 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 5 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 6 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 7 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

File: quantum/split_common/transport.c (hunk 8 above)
Comment (cwebster2) [reply quantum/split_common/transport.c]: I have implented this in my PR

**Comment by cwebster2** [comment]
I have incorporated all the suggested changes into my PR.

File: quantum/split_common/transport.c [hunk 9]
```diff
@@ -91,6 +101,23 @@ bool transport_master(matrix_row_t matrix[]) {
         }
     }
 #    endif
+    uint8_t real_mods = get_mods();
```
Comment (fauxpark) [inline quantum/split_common/transport.c:104]:
```suggestion

    uint8_t real_mods = get_mods();
```

File: quantum/split_common/transport.c [hunk 10]
```diff
@@ -251,6 +289,11 @@ bool transport_master(matrix_row_t matrix[]) {
     // Write wpm to slave
     serial_m2s_buffer.current_wpm = get_current_wpm();
 #    endif
+    serial_m2s_buffer.real_mods = get_mods();
```
Comment (fauxpark) [inline quantum/split_common/transport.c:292]:
```suggestion

    serial_m2s_buffer.real_mods = get_mods();
```

File: quantum/split_common/transport.c [hunk 11]
```diff
@@ -271,6 +314,11 @@ void transport_slave(matrix_row_t matrix[]) {
 #    ifdef WPM_ENABLE
     set_current_wpm(serial_m2s_buffer.current_wpm);
 #    endif
+    set_mods(serial_m2s_buffer.real_mods);
```
Comment (fauxpark) [inline quantum/split_common/transport.c:317]:
```suggestion

    set_mods(serial_m2s_buffer.real_mods);
```

File: quantum/split_common/transport.c [hunk 12]
```diff
@@ -91,6 +101,23 @@ bool transport_master(matrix_row_t matrix[]) {
...
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_WEAK_MODS_START, (void *)&weak_mods, sizeof(weak_mods), TIMEOUT) >= 0) {
+            i2c_buffer->weak_mods = weak_mods;
+    }
+
+    uint8_t oneshot_mods = get_mods();
+    if (oneshot_mods != i2c_buffer->oneshot_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_ONESHOT_MODS_START, (void *)&oneshot_mods, sizeof(oneshot_mods), TIMEOUT) >= 0) {
+            i2c_buffer->oneshot_mods = oneshot_mods;
+    }
     return true;
```
Comment (fauxpark) [inline quantum/split_common/transport.c:121]:
```suggestion
#endif

    return true;
```

File: quantum/split_common/transport.c [hunk 13]
```diff
@@ -91,6 +101,23 @@ bool transport_master(matrix_row_t matrix[]) {
...
+            i2c_buffer->real_mods = real_mods;
+    }
+
+    uint8_t weak_mods = get_weak_mods();
+    if (weak_mods != i2c_buffer->weak_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_WEAK_MODS_START, (void *)&weak_mods, sizeof(weak_mods), TIMEOUT) >= 0) {
+            i2c_buffer->weak_mods = weak_mods;
+    }
+
+    uint8_t oneshot_mods = get_mods();
```
Comment (fauxpark) [inline quantum/split_common/transport.c:116]:
```suggestion
#ifndef NO_ACTION_ONESHOT
    uint8_t oneshot_mods = get_oneshot_mods();
```

File: quantum/split_common/transport.c [hunk 14]
```diff
@@ -118,6 +145,12 @@ void transport_slave(matrix_row_t matrix[]) {
 #    ifdef WPM_ENABLE
     set_current_wpm(i2c_buffer->current_wpm);
 #    endif
+
+    set_mods(i2c_buffer->real_mods);
+    set_weak_mods(i2c_buffer->weak_mods);
+#ifndef NO_ACTION_ONESHOT
+    set_oneshot_mods(i2c_buffer->oneshot_mods);
+#endif
```
Comment (fauxpark) [inline quantum/split_common/transport.c:153]:
```suggestion
#    ifndef NO_ACTION_ONESHOT
    set_oneshot_mods(i2c_buffer->oneshot_mods);
#    endif
```

File: quantum/split_common/transport.c [hunk 15]
```diff
@@ -139,6 +172,11 @@ typedef struct _Serial_s2m_buffer_t {
 } Serial_s2m_buffer_t;
 
 typedef struct _Serial_m2s_buffer_t {
+    uint8_t real_mods;
+    uint8_t weak_mods;
+#ifndef NO_ACTION_ONESHOT
+    uint8_t oneshot_mods;
+#endif
```
Comment (fauxpark) [inline quantum/split_common/transport.c:179]:
```suggestion
#    ifndef NO_ACTION_ONESHOT
    uint8_t oneshot_mods;
#    endif
```

File: quantum/split_common/transport.c [hunk 16]
```diff
@@ -251,6 +289,11 @@ bool transport_master(matrix_row_t matrix[]) {
     // Write wpm to slave
     serial_m2s_buffer.current_wpm = get_current_wpm();
 #    endif
+    serial_m2s_buffer.real_mods = get_mods();
+    serial_m2s_buffer.weak_mods = get_weak_mods();
+#ifndef NO_ACTION_ONESHOT
+    serial_m2s_buffer.oneshot_mods = get_oneshot_mods();
+#endif
```
Comment (fauxpark) [inline quantum/split_common/transport.c:296]:
```suggestion
#    ifndef NO_ACTION_ONESHOT
    serial_m2s_buffer.oneshot_mods = get_oneshot_mods();
#    endif
```

File: quantum/split_common/transport.c [hunk 17]
```diff
@@ -271,6 +314,11 @@ void transport_slave(matrix_row_t matrix[]) {
 #    ifdef WPM_ENABLE
     set_current_wpm(serial_m2s_buffer.current_wpm);
 #    endif
+    set_mods(serial_m2s_buffer.real_mods);
+    set_weak_mods(serial_m2s_buffer.weak_mods);
+#ifndef NO_ACTION_ONESHOT
+    set_oneshot_mods(serial_m2s_buffer.oneshot_mods);
+#endif
```
Comment (fauxpark) [inline quantum/split_common/transport.c:321]:
```suggestion
#    ifndef NO_ACTION_ONESHOT
    set_oneshot_mods(serial_m2s_buffer.oneshot_mods);
#    endif
```

File: quantum/split_common/transport.c [hunk 18]
```diff
@@ -28,7 +28,14 @@ static pin_t encoders_pad[] = ENCODERS_PAD_A;
 
 typedef struct _I2C_slave_buffer_t {
     matrix_row_t smatrix[ROWS_PER_HAND];
+    uint8_t real_mods;
+    uint8_t weak_mods;
+#ifndef NO_ACTION_ONESHOT
+    uint8_t oneshot_mods;
+#endif
```
Comment (fauxpark) [inline quantum/split_common/transport.c:35]:
```suggestion
#    ifndef NO_ACTION_ONESHOT
    uint8_t oneshot_mods;
#    endif
```

File: quantum/split_common/transport.c [hunk 19]
```diff
@@ -45,6 +52,9 @@ static I2C_slave_buffer_t *const i2c_buffer = (I2C_slave_buffer_t *)i2c_slave_re
 #    define I2C_BACKLIGHT_START offsetof(I2C_slave_buffer_t, backlight_level)
 #    define I2C_RGB_START offsetof(I2C_slave_buffer_t, rgblight_sync)
 #    define I2C_KEYMAP_START offsetof(I2C_slave_buffer_t, smatrix)
+#    define I2C_REAL_MODS_START offsetof(I2C_slave_buffer_t, real_mods)
+#    define I2C_WEAK_MODS_START offsetof(I2C_slave_buffer_t, weak_mods)
+#    define I2C_ONESHOT_MODS_START offsetof(I2C_slave_buffer_t, oneshot_mods)
```
Comment (fauxpark) [inline quantum/split_common/transport.c:57]:
```suggestion
#    define I2C_KEYMAP_START offsetof(I2C_slave_buffer_t, smatrix)
#    define I2C_REAL_MODS_START offsetof(I2C_slave_buffer_t, real_mods)
#    define I2C_WEAK_MODS_START offsetof(I2C_slave_buffer_t, weak_mods)
#    define I2C_ONESHOT_MODS_START offsetof(I2C_slave_buffer_t, oneshot_mods)
#    define I2C_BACKLIGHT_START offsetof(I2C_slave_buffer_t, backlight_level)
#    define I2C_RGB_START offsetof(I2C_slave_buffer_t, rgblight_sync)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Some slight cosmetic fixes.

**Comment by cwebster2** [comment]
I think i've now addressed the above comments.

File: quantum/split_common/transport.c [hunk 20]
```diff
@@ -91,6 +103,27 @@ bool transport_master(matrix_row_t matrix[]) {
...
+    if (weak_mods != i2c_buffer->weak_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_WEAK_MODS_START, (void *)&weak_mods, sizeof(weak_mods), TIMEOUT) >= 0) {
+            i2c_buffer->weak_mods = weak_mods;
+    }
+
+#    ifndef NO_ACTION_ONESHOT
+    uint8_t oneshot_mods = get_oneshot_mods();
+    if (oneshot_mods != i2c_buffer->oneshot_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_ONESHOT_MODS_START, (void *)&oneshot_mods, sizeof(oneshot_mods), TIMEOUT) >= 0) {
+            i2c_buffer->oneshot_mods = oneshot_mods;
```
Comment (fauxpark) [inline quantum/split_common/transport.c:123]:
```suggestion
            i2c_buffer->oneshot_mods = oneshot_mods;
        }
```

File: quantum/split_common/transport.c [hunk 21]
```diff
@@ -91,6 +103,27 @@ bool transport_master(matrix_row_t matrix[]) {
...
+    uint8_t real_mods = get_mods();
+    if (real_mods != i2c_buffer->real_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_REAL_MODS_START, (void *)&real_mods, sizeof(real_mods), TIMEOUT) >= 0) {
+            i2c_buffer->real_mods = real_mods;
+    }
+
+    uint8_t weak_mods = get_weak_mods();
+    if (weak_mods != i2c_buffer->weak_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_WEAK_MODS_START, (void *)&weak_mods, sizeof(weak_mods), TIMEOUT) >= 0) {
+            i2c_buffer->weak_mods = weak_mods;
```
Comment (fauxpark) [inline quantum/split_common/transport.c:116]:
```suggestion
            i2c_buffer->weak_mods = weak_mods;
        }
```

File: quantum/split_common/transport.c [hunk 22]
```diff
@@ -91,6 +103,27 @@ bool transport_master(matrix_row_t matrix[]) {
         }
     }
 #    endif
+
+    uint8_t real_mods = get_mods();
+    if (real_mods != i2c_buffer->real_mods) {
+        if (i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_REAL_MODS_START, (void *)&real_mods, sizeof(real_mods), TIMEOUT) >= 0) {
+            i2c_buffer->real_mods = real_mods;
```
Comment (fauxpark) [inline quantum/split_common/transport.c:110]:
```suggestion
            i2c_buffer->real_mods = real_mods;
        }
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/split_common/transport.c [hunk 23]
```diff
@@ -42,9 +49,14 @@ typedef struct _I2C_slave_buffer_t {
 
 static I2C_slave_buffer_t *const i2c_buffer = (I2C_slave_buffer_t *)i2c_slave_reg;
 
+#    define I2C_KEYMAP_START offsetof(I2C_slave_buffer_t, smatrix)
+#    define I2C_REAL_MODS_START offsetof(I2C_slave_buffer_t, real_mods)
+#    define I2C_WEAK_MODS_START offsetof(I2C_slave_buffer_t, weak_mods)
+#    ifndef NO_ACTION_ONESHOT
+#    define I2C_ONESHOT_MODS_START offsetof(I2C_slave_buffer_t, oneshot_mods)
+#    endif
```
Comment (fauxpark) [inline quantum/split_common/transport.c:57]:
```suggestion
#    define I2C_ONESHOT_MODS_START offsetof(I2C_slave_buffer_t, oneshot_mods)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Review by zvecr — CHANGES_REQUESTED** [review]
Given this will change the scan rate a fair bit, this change should be optional, opt-in rather than out. That way boards which have zero reason to use the data, no oled etc, can remain as fast as they used to be.

**Comment by cwebster2** [comment]
> Given this will change the scan rate a fair bit, this change should be optional, opt-in rather than out. That way boards which have zero reason to use the data, no oled etc, can remain as fast as they used to be.

The added code is now behind a define to make it opt-in.

**Comment by drashna** [comment]
Also, could you add documentation for this.  It doesn't need to be extensive, but it should be documented.  
https://github.com/qmk/qmk_firmware/blob/master/docs/feature_split_keyboard.md

**Comment by cwebster2** [comment]
> Also, could you add documentation for this. It doesn't need to be extensive, but it should be documented.
> https://github.com/qmk/qmk_firmware/blob/master/docs/feature_split_keyboard.md

I have documented the define under communication options on this page and made mention that it of cosmetic use and may impact the matrix scan rate if enabled per zvecr's earlier comment in this PR.

**Review by drashna — APPROVED** [review]

**Comment by fauxpark** [comment]
@cwebster2 looks like this needs a rebase, sorry.

**Comment by cwebster2** [comment]
I will take care of that later today.

**Comment by cwebster2** [comment]
@fauxpark I have rebased my branch against qmk/develop

**Comment by cwebster2** [comment]
@zvecr this has probably fallen off the radar, but wondering if your change request has been satisfied and it this is ready to merge.

**Comment by cwebster2** [comment]
I've rebased this PR against develop yet again, but squashed it into a single commit this time.

Out of curiosity, what workflow is in use in this branch that results in this frequent history rewriting?

**Pull request merged**